Event with blocking

i-am-tom opened the pull request:
The `withBlocking` function uses `fix` to allow us to block inputs based on values discovered later in the process. A lot of the insight here comes from @monoidmusician, to whom I'm eternally grateful.

Commit:
Add `gateBy` and `gate` to `FRP.Event.Time`

Previously, the functions only existed for `Behavior`. This PR adds them
to `Event` as well.

Commit:
Fix comments from review

Put the functions in the wrong place. Oops. This commit moves the
`gate`-related functions from `FRP.Event.Time` to `FRP.Event.Class`, and
exports them.

Commit:
Add withBlocking to FRP.Event.Time

The `withBlocking` function allows the user to add dynamic debouncing by
providing some delay period-calculating process. `debounce` simply fixes the
period to a given number of seconds.

paf31 commented on src/FRP/Event/Time.purs:
```diff
@@ -15,3 +17,38 @@ foreign import animationFrame :: Event Unit
+    stamped = withTime event
+
+-- | On each event, ignore subsequent events for a given number of milliseconds.
+debounce :: forall a. Number -> Event a -> Event a
```
I think `debounce` usually refers to a similar behavior, but where the final event does actually fire, after the blocking period, right?

i-am-tom replied:
Hmm, ok. Might need to do some applicative wizardry to get this working, then. I'll get to it tonight :)

paf31 commented on src/FRP/Event/Time.purs:
```diff
@@ -15,3 +17,38 @@ foreign import animationFrame :: Event Unit
+        comparison a b = a < b.time
+
+        unblocked :: Event { time :: Number, value :: a }
+        unblocked = gateBy comparison expiries stamped
```
Could you please inline `comparison` here? I think it would make it slightly easier to read.

i-am-tom replied:
No worries!

paf31 commented on src/FRP/Event/Time.purs:
```diff
@@ -15,3 +17,38 @@ foreign import animationFrame :: Event Unit
+        processed = process allowed
+
+        expiries :: Event Number
+        expiries = pure 0.0 <|>
```
Zero seems like an odd choice here. What about representing the expiry as a predicate on times instead? Then you could start with `const true` here.

i-am-tom replied:
Am I then saying that each loop adds a new predicate on whether times are above its expiry? Is there a perf concern here? (I'll do some profiling over long time periods)

MonoidMusician replied:
Yeah, I agree, 0.0 isn't the best ... What about using `Maybe` instead?

paf31 replied:
Sounds good.

paf31 commented on src/FRP/Event/Time.purs:
```diff
@@ -15,3 +17,38 @@ foreign import animationFrame :: Event Unit
+     (Event a -> Event { period :: Number, value :: b })
+  -> Event a
+  -> Event b
+withBlocking process event
```
Since this generalizes `debounce` (or whatever we end up calling it), how about naming this something like `debounceWith`?

i-am-tom replied:
Yep, no worries :D

Commit:
Implement code review changes

As per @paf31's code review, there is now an "initial" value for
`gateBy`, which is `false` for `gate`. This behavior has been documented
in both places, and `withBlocking` has been updated accordingly (the
"initial time" is the epoch).

i-am-tom commented:
Closing to consolidate with the other PR
